### ISSUE TITLE
linux coming out of full screen resize issues

### ISSUE THREAD

**Issue by yukapuka123** [glfw/glfw#671]
I tested my program in windows and in linux 
in windows it works fine 
in linux it does strange things 
it seems to leave the created window and also exists the screen by itself and does not resize the screen to the previous size or chosen desktop size

for example i created a window that is of size 640*480 , while my desktop size if 1920*1080 
once i launch the program , it works for maybe a second or while the mouse is within the created window, and once you leave the created window size it scrolls past it and exits the created window context , however it also doesnt return you to your normal desktop size

**Comment by tombsar** [comment]
It sounds as though your program is crashing, and therefore not cleaning up properly. Can you determine what is causing the program to terminate, possibly with gdb?

**Comment by yukapuka123** [comment]
how can it be crashing if the SAME program is working in windows, but in linux giving the problem that it is giving 

if it were crashing it would also crash in windows

**Comment by slimsag** [comment]
@yukapuka123 different platforms behave in different ways. It's not strange to think that the program might run fine on Windows but not on Linux. Can you provide a sample code that reproduces the issue you are seeing on Linux?

**Comment by yukapuka123** [comment]
i have not written anything complicated , just created a window in FULL SCREEN mode(with the option for glfwGetPrimaryMonitor() in the appropriate perimeter of the function try it yourself and see what you 

create  window that is full-screen in linux of size 640 * 480 and tell me what you get, because THAT is all i created, nothing else  

GLFWwindow* window = glfwCreateWindow(640, 480, "My Title", glfwGetPrimaryMonitor(), NULL);

as i said it works in windows perfectly, compiles and runs and does what it is supposed to of, but NOT in linux , in linux it has a mind of its own

**Comment by yukapuka123** [comment]
#include <GL/glew.h> // include GLEW and new version of GL on Windows
#include <GLFW/glfw3.h> // GLFW helper library
#include <stdio.h>
#include <glm/glm.hpp>
 
void key_callback(GLFWwindow* window, int key, int scancode, int action, int mods)
{
    if (key == GLFW_KEY_ESCAPE && action == GLFW_PRESS)
        glfwSetWindowShouldClose(window, GL_TRUE);
}
static void cursor_position_callback(GLFWwindow* window, double xpos, double ypos)
{
   printf("xpos is %lf, ypos is %lf\n", xpos, ypos);
}

int main(){
  
   if (!glfwInit ()) {
      fprintf (stderr, "ERROR: could not start GLFW3\n");
      return 1;
   } 
 
    
 
   glfwWindowHint (GLFW_CONTEXT_VERSION_MINOR, 2);
   glfwWindowHint (GLFW_CONTEXT_VERSION_MAJOR, 3);
   glfwWindowHint (GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
 
   GLFWwindow* window = glfwCreateWindow (1280, 720, "Hello Triangle",  NULL, NULL);
 
   if (!window) {
      fprintf (stderr, "ERROR: could not open window with GLFW3\n");
      glfwTerminate();
   return 1;
   }
 
   glfwMakeContextCurrent (window);
 
    // start GLEW extension handler
   glewExperimental = GL_TRUE;
   GLuint err = glewInit();
   if(err!=GLEW_OK){
      fprintf(stderr, "Error: %s\n", glewGetErrorString(err));
      return 1;
   }
 
   glfwSetKeyCallback(window, key_callback);
   glfwSetCursorPosCallback(window, cursor_position_callback);
   
 
   while(!glfwWindowShouldClose(window)){
      
      // put the stuff we've been drawing onto the display
      glfwSwapBuffers (window);
      // update other events like input handling 
      glfwPollEvents ();
   }
 
   glfwTerminate();
    
}

**Comment by Spartan322** [comment]
@yukapuka123 two things, one, you might have wanted to use pastebin for stuff like that, and second, GLFW does not run the same code for Windows and Linux, because it can't run the same code, two different OSes handle windows differently, and this is where behavior can change, but the objective is to mimic behavior as close as practical.

**Comment by yukapuka123** [comment]
the GLFW guide specifically says that it is OS independent and that the code will run in all OS's with the same results 

surely if they created the API they tested it and wouldn't be just saying that and hoping it works

**Comment by Spartan322** [comment]
except you can't actually do that, we can't run windows the same in ever OS, we abstract the window system, but its still there, we just try to make them work as close as possible (an example is we can't do certain things on Mac like we can on Windows or Linux because they do not in fact the use the same code). If I have to show you where the Linux, Mac, and Windows code all diverge, I can.

**Comment by yukapuka123** [comment]
nevertheless, why doesn't this work? is it something i am missing in the code? or is it something unfixable? and what is a possible solution or at least work around?

**Comment by yukapuka123** [comment]
all of you wanted to see the code, i show you the code and suddenly everyone stops being smart, before that everyone was smart and was explaining this and that, so what now? all your "genius" left you all? nobody has a solution to such a simple thing?

**Comment by Spartan322** [comment]
don't hijack issues and act mean, some of us are working on finding a solution, and tbh, this is probably not a simple problem. Don't forget that not everyone is sleeping and working on the same clock.

**Comment by yukapuka123** [comment]
so your one of the developers of glfw? i thought you were some random 

okay fair enough 

anyway i think this is an issue with re-sizing the screen, or maybe it could be an issue with directly with X11 , i heard it is terrible implementation and everyone is trying to move to wayland

**Comment by Spartan322** [comment]
Also, I'm not a dev for glfw, but I prefer to help it with progress, so I come by every so often (when I can).

Anyway, I'll try troubleshooting on a virtual machine tomorrow, its rather late for me (unless someone else can describe the problem by then). Could you tell me the version of X11 that your using?

**Comment by tombsar** [comment]
To avoid any confusion, glfw is intended to work on various Linux systems, and is tested and actively used by many people on Linux (including myself). Therefore the issue you are experiencing is more subtle than an obvious mistake in the library that we can find without input from you.

I asked if your program is terminating abnormally because in your original post you mentioned that the window closes despite your code not telling it to. In general it is not unusual for code that works on one OS and compiler to crash on a different system, so I do not think this is an unreasonable question to ask.

Looking at the code you posted, I think I may see the problem. If GLEW fails to initialize, you simply return from main without terminating glfw. Can you please check if this is happening?

**Comment by elmindreda** [comment]
@yukapuka123 Your attitude stinks. The people you're talking to are volunteers donating their spare time to help others. You could show a little gratitude.

This is a possible duplicate of #474, which I still can't reproduce.

**Comment by yukapuka123** [comment]
do you want me to make a screen capture video?

**Comment by Spartan322** [comment]
@yukapuka123 If it can't be reproduced by enough people, there is almost no way for us to figure out the problem (recording it won't really help, we need to be able to test on our own, with our own solutions too). I could try to reproduce on a VM, but I need to know the version of X11 your using.

**Comment by andrewrk** [comment]
@Spartan322 there is one way to try to debug something tricky when only one person can reproduce the problem - they can generate a core dump of the crash. This is a huge file that contains all the memory in the virtual address space of the program at the time of the crash. This along with causing the crash to happen with GLFW in a debug build would result in a core dump that a developer could use to inspect the state of the program just as if the developer was able to cause the crash on their own system.

However, in this specific instance, it is not clear to me what exactly is the expected behavior vs actual behavior. It sounds like perhaps @yukapuka123 misunderstands how X11 and/or their window manager works.

**Comment by Spartan322** [comment]
@andrewrk Yeah, I realize that (in fact, a dump would be really helpful right now), and if its the case of expected vs. reality, there is no fixing that, however, from what I've read, I'm not sure the result that is suppose to be reality, but I can only guess.

Plus, more testing is usually a good thing, so I guess it doesn't matter too much

But if @yukapuka123 wants help with solving (if it is able to/should be solved), then the actual OS info would be needed

**Comment by tombsar** [comment]
We haven't even established that the program is crashing yet...

@yukapuka123 please can you check whether your program is terminating (normally or abnormally) when the issue occurs, and whether or not GLEW is initializing correctly. I cannot help further until you answer these questions.

**Comment by yukapuka123** [comment]
https://youtu.be/SQC9I2qmHzc

here is a screen capture of the video(its private so fear not i am not exposing this to youtube)

**Comment by Spartan322** [comment]
Is it the background not disappearing? What is the expect behavior here? And what is actually happening contrary to that?

**Comment by yukapuka123** [comment]
so as you can see the first time i run it(my resolution is in 1920 * 1080) and the resolution doesnt actually change , the full 1920* 1080 is still accessible 

and then when you exit the program window(glfw) the resolution that it returns to is 640 * 480 as you can see that it has been set to in the display settings and hence i have to change it back

**Comment by yukapuka123** [comment]
i didnt look through your window creation for linux process in your code , but i assume your using Xlib and GLX 

but in looking at GLX and Xlib around on the internet for creation window context

i found that people where having a similar problem with full screen workarounds 

and they basically solved that problem by overriding the window manager , here is an example for a solution somebody posted in a forum for overriding x  window manager(which was causing problem to people) and i noticed you use Xinerama, Xrandr so maybe that is where the problem lies?

Hi,
i had the same problem and browsed through tons of documentation...
Now I think that I have found the way to do it, and its quite simple. You simply have to set the "override-redirect"-flag to True and the window size to screen resolution. The code sample below lacks any error checking and features to switch screen resolution or similar, but i gives me a fullscreen-window without decorations :-)

Here comes the code:

/* attribute list for glx-visual */
static int attrList[] = { GLX_RGBA, GLX_DOUBLEBUFFER, 
GLX_RED_SIZE, 1, 
GLX_GREEN_SIZE, 1, 
GLX_BLUE_SIZE, 1, 
None };

int main(int argc, char **argv)
{
Display *dpy;
int screen;
XVisualInfo *vi;
Colormap cmap;
Window win;
GLXContext ctx;
XEvent event;
XSetWindowAttributes attr;
int dpy_width, dpy_height;

/* get a connection */
dpy = XOpenDisplay(0);
screen = DefaultScreen(dpy);
/* get the screen resolution */
dpy_width = DisplayWidth (dpy, screen);
dpy_height = DisplayHeight (dpy, screen);

/* get a visual */
vi = glxChooseVisual (dpy, screen, attrList);

/* create a GLX context */
ctx = glXCreateContext(dpy, vi, 0, GL_TRUE);

/* create a color map */
cmap = XCreateColormap(dpy, RootWindow(dpy, vi->screen),
vi->visual, AllocNone);

/* create a window */
attr.colormap = cmap;
attr.border_pixel = 0;
/* prevent the window-manager from decorating our window */
attr.override_redirect = True;
attr.event_mask = ExposureMask | KeyPressMask | ButtonPressMask |
StructureNotifyMask;
win = XCreateWindow(dpy, RootWindow(dpy, vi->screen), 0, 0, dpy_width,
dpy_height, 0, vi->depth, InputOutput, vi->visual, CWBorderPixel|CWColormap|CWEventMask|CWOverrideRed irect, &attr);
/* display our window in front of everything else */
XMapRaised(dpy, win);
....
}

I hope this is what everyone wanted and that it works not only for me...

cheers and happy coding

bxe Miq

**Comment by Spartan322** [comment]
I though we already did this, I'm pretty sure GLFW already handles this case

**Comment by yukapuka123** [comment]
what do you mean? as in the solution provided here is already incorporated into GLFW?

**Comment by Spartan322** [comment]
Yes, I believe we already have this incorporated, let me take a look in x11_window.c

**Comment by yukapuka123** [comment]
"2.3.2
Graphics Reset Recovery
Certain events can result in a reset of the GL context. After such an event, it is
referred to as a lost context and is unusable for almost all purposes. Recovery re-
quires creating a new context and recreating all relevant state from the lost context.
The current status of the graphics reset state is returned by
enum GetGraphicsResetStatus( void );"

from the specification 

could have the context been reset? is there error checking on this issue in GLFW?

**Comment by yukapuka123** [comment]
@Spartan322 just a quick question about glfw window creation , once a window is created 
i.e glfwCreateWindow (); it instantly pops up on the screen right? 

because when i am drawing without using a loop 
like this :

(after everything is correctly initialized and linked and compiled and binded etc )
instead of a loop i just go like this:
glClear (GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
glDrawArrays (GL_TRIANGLES, 0, 3);
glfwSwapBuffers(window);
sleep(3)

the window pops up and stays there on scree, however it doesnt seem to swap buffer, i get a screen as if i didnt draw anything(i.e transparent) 

is the reason that the application window is running faster then opengl? because when i add 2-3 (between the draw operations)swapBuffers i then see what i expected to see 

the code is messy but nevertheless if you want to see it i can paste it

**Comment by Spartan322** [comment]
I don't think you aren't really drawing, you are just knocking out the thread after trying to create an image

**Comment by yukapuka123** [comment]
if i actually add more glfwswapbuffers... it does show

**Comment by yukapuka123** [comment]
glfwSwapBuffers(window);
glClear (GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
glDrawArrays (GL_TRIANGLES, 0, 3);
glfwSwapBuffers(window);
sleep(1);

like this it will actually swap the buffer and show the drawing and i was just wondering is it maybe that the gl commands dont finish in time? or something else

i tried glFlush() and it didnt make a difference

**Comment by yukapuka123** [comment]
it seems that one glfwSwapBuffer() command is not enough to make the first swap in this stance 
however if i add sleep(1) before it does work in this way 
sleep(1);
glClear (GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
glDrawArrays (GL_TRIANGLES, 0, 3);
glfwSwapBuffers(window);
sleep(1);

**Comment by Spartan322** [comment]
dude, just edit the comment, people don't like it when they get wasteful notifications like this

**Comment by yukapuka123** [comment]
what do you mean edit the comment?

**Comment by Spartan322** [comment]
there is a little pencil in github comments, click that and edit the comment instead of doing multiple posts in a row

**Comment by yukapuka123** [comment]
okay ive done what you've asked, anyway any idea what the issue is? why it works when you put it to sleep for even just 1 second, but otherwise unless you swap the buffer a few times it doesnt show anything?

**Comment by Spartan322** [comment]
does this count as issue hijacking? either way, its obviously due to how the buffers operate, I'm not sure if that can change.

**Comment by yukapuka123** [comment]
http://stackoverflow.com/questions/26162745/glxswapbuffers-blocks-until-vblank

i think i found the answer to the problem, it is to do with glxSwapBuffers

**Comment by shurcooL** [comment]
> people don't like it when they get wasteful notifications like this

For what it's worth, if you're using GitHub's notification center, you don't get extra notifications per-comment; the entire thread shows up as "having new comments".

![image](https://cloud.githubusercontent.com/assets/1924134/12008707/9082aee2-ac04-11e5-852b-f9f3921e8f4e.png)

That said, it is of course nicer to edit existing comments when there are typos/minor issues to fix, rather than creating a new one. The edit button (pencil icon) in the top right corner is exactly for that.

**Comment by tombsar** [comment]
@yukapuka123 On X11 a lot of things happen asynchronously, so the window and context are probably not created in time for your first set of draw commands. Try putting `glfwWaitEvents` before you draw (first time only).
